linux/tests/async_socket_linux_int/async_socket_linux_int.c

```diff
@@ -0,0 +1,520 @@
+    wake_by_address_single(thread_counter);
+}
+
+static void on_receive_complete(void* context, ASYNC_SOCKET_RECEIVE_RESULT receive_result, uint32_t bytes_received)
+{
+    volatile_atomic int32_t* thread_counter = (volatile_atomic int32_t*)context;
+
+    ASSERT_ARE_EQUAL(ASYNC_SOCKET_RECEIVE_RESULT, ASYNC_SOCKET_RECEIVE_OK, receive_result);
+    (void)interlocked_increment(thread_counter);
+    wake_by_address_single(thread_counter);
+}
+
+static void on_receive_and_accumulate_complete(void* context, ASYNC_SOCKET_RECEIVE_RESULT receive_result, uint32_t bytes_received)
+{
+    volatile_atomic int32_t* accumulator = (volatile_atomic int32_t*)context;
+
+    ASSERT_ARE_EQUAL(ASYNC_SOCKET_RECEIVE_RESULT, ASYNC_SOCKET_RECEIVE_OK, receive_result);
+    (void)interlocked_add(accumulator, bytes_received);
+    wake_by_address_single(accumulator);
+}
+
+static void on_receive_complete_with_error(void* context, ASYNC_SOCKET_RECEIVE_RESULT receive_result, uint32_t bytes_received)
+{
+    volatile_atomic int32_t* thread_counter = (volatile_atomic int32_t*)context;
+
+    ASSERT_ARE_EQUAL(ASYNC_SOCKET_RECEIVE_RESULT, ASYNC_SOCKET_RECEIVE_ABANDONED, receive_result);
+    (void)interlocked_increment(thread_counter);
+    wake_by_address_single(thread_counter);
+}
+
+static void set_nonblocking(SOCKET_HANDLE socket)
+{
+    int opts = fcntl(socket, F_GETFL);
+    ASSERT_IS_TRUE(opts >= 0, "Failure getting socket option");
+
+    opts = fcntl(socket, F_SETFL, opts|O_NONBLOCK);
+    ASSERT_IS_TRUE(opts >= 0, "Failure setting socket option");
+}
+
+static void setup_server_socket(SOCKET_HANDLE* listen_socket)
+{
+    // create a listening socket
+    *listen_socket = socket(AF_INET, SOCK_STREAM, 0);
+    ASSERT_ARE_NOT_EQUAL(int, INVALID_SOCKET, *listen_socket);
+
+    const int enable = 1;
+    ASSERT_ARE_EQUAL(int, 0, setsockopt(*listen_socket, SOL_SOCKET, SO_REUSEADDR, &enable, sizeof(int)));
+
+    int bind_res;
+    uint32_t counter = 0;
+    // The gate machines fails on bind due to 3 different
+    // Process running, so we have to figure out the bind
+    do
+    {
+        struct sockaddr_in service;
+
+        service.sin_family = AF_INET;
+        service.sin_port = htons(g_port_num);
+        service.sin_addr.s_addr = htonl(INADDR_ANY);
+
+        bind_res = bind(*listen_socket, (struct sockaddr*)&service, sizeof(service));
+        if (bind_res != 0)
+        {
+            g_port_num++;
+        }
+        counter++;
+    } while (bind_res != 0 && counter < 10);
+    ASSERT_ARE_EQUAL(int, 0, bind_res, "Failure attempting to bind (%d) to socket %d error (%d): %s", bind_res, g_port_num, errno, strerror(errno));
+
+    // set it to async IO
+    set_nonblocking(*listen_socket);
+
+    // start listening
+    ASSERT_ARE_EQUAL(int, 0, listen(*listen_socket, SOMAXCONN), "Failure on listen socket");
+
+}
+
+static void setup_client_sockets(int port_num, SOCKET_HANDLE* client_socket, SOCKET_HANDLE* listen_socket, SOCKET_HANDLE* accept_socket)
```

Comment:
>setup_client_sockets [](http://example.com/codeflow?start=12&length=20)

Sounds like it is setting up just one client socket.
But also seems like it is setting up both client and server sockets.

So maybe it should be named setup_seockets or setup_test_sockets